gear-programs/vft-treasury/src/services/mod.rs

```diff
@@ -0,0 +1,433 @@
+    pub ethereum_event_client: ActorId,
+    pub config: Config,
+}
+
+impl InitConfig {
+    pub fn new(
+        receiver_contract_address: H160,
+        gear_bridge_builtin: ActorId,
+        ethereum_event_client: ActorId,
+        config: Config,
+    ) -> Self {
+        Self {
+            receiver_contract_address,
+            gear_bridge_builtin,
+            ethereum_event_client,
+            config,
+        }
+    }
+}
+impl<T> VftTreasury<T>
+where
+    T: ExecContext,
+{
+    pub fn seed(config: InitConfig, exec_context: T) {
+        unsafe {
+            DATA = Some(VftTreasuryData {
+                receiver_contract_address: config.receiver_contract_address,
+                gear_bridge_builtin: config.gear_bridge_builtin,
+                ethereum_event_client: config.ethereum_event_client,
+                admin: exec_context.actor_id(),
+                vara_eth_mapping: Vec::new(),
+            });
+            CONFIG = Some(config.config);
+            MSG_TRACKER = Some(MessageTracker::default());
+        }
+    }
+
+    pub fn new(exec_context: T) -> Self {
+        Self { exec_context }
+    }
+
+    fn data(&self) -> &VftTreasuryData {
+        unsafe { DATA.as_ref().expect("VftTreasury::seed() must be called") }
+    }
+
+    fn data_mut(&mut self) -> &mut VftTreasuryData {
+        unsafe { DATA.as_mut().expect("VftTreasury::seed() must be called") }
+    }
+
+    fn config(&self) -> &Config {
+        unsafe {
+            CONFIG
+                .as_ref()
+                .expect("VftTreasury::seed() must be invoked")
+        }
+    }
+
+    fn get_eth_token_id(&self, vara_token_id: &ActorId) -> Result<H160, Error> {
+        self.data()
+            .vara_eth_mapping
+            .iter()
+            .find(|(vara, _)| vara_token_id == vara)
+            .map(|(_, eth)| *eth)
+            .ok_or(Error::NoCorrespondingEthAddress)
+    }
+
+    fn get_vara_token_id(&self, eth_token_id: &H160) -> Result<ActorId, Error> {
+        self.data()
+            .vara_eth_mapping
+            .iter()
+            .find(|(_, eth)| eth_token_id == eth)
+            .map(|(vara, _)| *vara)
+            .ok_or(Error::NoCorrespondingEthAddress)
+    }
+}
+
+#[derive(Encode, Decode, TypeInfo)]
+pub enum VftTreasuryEvents {
+    Deposit {
+        from: ActorId,
+        to: H160,
+        token: ActorId,
+        amount: U256,
+    },
+    Withdraw {
+        receiver: ActorId,
+
+        token: ActorId,
+        amount: U256,
+    },
+}
+
+#[service(events = VftTreasuryEvents)]
+impl<T> VftTreasury<T>
+where
+    T: ExecContext,
+{
+    pub fn ensure_admin(&mut self) -> Result<(), Error> {
+        if self.data().admin != self.exec_context.actor_id() {
+            return Err(Error::NotAdmin);
+        }
+
+        Ok(())
+    }
+
+    pub fn update_config(&mut self, config: Config) -> Result<(), Error> {
+        if self.data().admin != self.exec_context.actor_id() {
+            return Err(Error::NotAdmin);
+        }
```

Comment:
You're still using this check everywhere but it can be replaced with `ensure_admin`